### ISSUE TITLE
Dockerd eats 20GB of RAM

### ISSUE THREAD

**Issue by DmitriyMV** [moby/moby#32711]
Dockerd eats up a huge chunk of memory when there are a lot of containers are starting and stopping on the same server.

**Steps to reproduce:**
`while :; do docker run --rm busybox /bin/false; done`

pprof output:
[pprof001.zip](https://github.com/moby/moby/files/933900/pprof001.zip)

**Output of `docker version`:**
```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**
```
Containers: 17
 Running: 17
 Paused: 0
 Stopped: 0
Images: 38
Server Version: 1.12.1
Storage Driver: overlay2
 Backing Filesystem: xfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local local-persist
 Network: host overlay bridge null macvlan
Swarm: active
 NodeID: oojd1ntn6u5javok5gdb5dtmn
 Is Manager: true
 ClusterID: 2wjx7mbqdze2cp25qvfpvwg0l
 Managers: 4
 Nodes: 4
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.17.2
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-34-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 24
Total Memory: 94.41 GiB
Name: ds1
ID: 6CA2:ROZW:SCEA:EDTR:SZXT:47N4:GA5O:EAAD:6OVN:PK6U:NGQ4:23E6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Comment by fortinj66** [comment]
Very similar issue here.  dockerd eats ups all the memory on the servers over the course of several days and docker and openshift origin have to be restarted.  Note that stopping all the containers has no effect on the memory used by dockerd

1081 root      20   0 20.756g 0.014t  15612 S   9.9 40.4 561:59.40 dockerd-current

[johnf@os-node-s01 ~]$ docker info
Containers: 44
 Running: 44
 Paused: 0
 Stopped: 0
Images: 27
Server Version: 1.12.5
Storage Driver: devicemapper
 Pool Name: docker--vg-docker--pool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 21.33 GB
 Data Space Total: 44.3 GB
 Data Space Available: 22.98 GB
 Metadata Space Used: 8.372 MB
 Metadata Space Total: 54.53 MB
 Metadata Space Available: 46.15 MB
 Thin Pool Minimum Free Space: 4.43 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: systemd
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: inactive
Runtimes: runc docker-runc
Default Runtime: docker-runc
Security Options: seccomp selinux
Kernel Version: 3.10.0-514.10.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
Number of Docker Hooks: 2
CPUs: 8
Total Memory: 35.18 GiB
Name: os-node-s01
ID: 7ZD3:FQ3S:SUWP:UY5L:DA22:DNUL:RLMJ:ZTUW:DH3X:KM46:NKBU:LPT6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 172.30.0.0/16
 127.0.0.0/8
Registries: docker.io (secure)

**Comment by pascalandy** [comment]
Never had this issue. Also 1.12 is getting old. 
Please update to latest daemon.

**Comment by cpuguy83** [comment]
If you can enable debug mode and grab some pprof details we can look into this, unfortunately as is there is no way to debug it.

What's your logging story, are you using a remote logger? Are logging buffers filling up b/c the log connection is down/slow?

To enable debug you can add `"debug": true` to `/etc/docker/daemon.json` then send `SIGHUP` to dockerd. If the file doesn't exist, create it... must be valid json.

Then you can hit the docker daemon with a `GET` request at `/debug/pprof/heap`. This will spit out a memory profile that you can import into `go tool pprof --inuse_space $(which dockerd) <path to memory profile>`... in addition it may be helpful to run a 2nd time with `--alloc_space`.

**Comment by fortinj66** [comment]
I have to stay with the Centos/RedHat/OpenShift supported versions. Once 13 is recommended I'll be happy to upgrade. 

--John

**Comment by fortinj66** [comment]
I'll give this a try as soon as I can...

--John

**Comment by fortinj66** [comment]
I'm using the journald log driver.  I’ve also tried the json driver and log rotation but it did not seem to help matters…

—John

**Comment by redbaron** [comment]
can't reproduce on 17.04.0-ce with btrfs storage driver

**Comment by unclejack** [comment]
I've tried to reproduce this issue using Docker 1.12.6 and 10000 containers. Memory usage didn't grow above 60 MB. Memory was being returned to the OS as expected.

It seems likely that the culprit wasn't `while :; do docker run --rm busybox /bin/false; done`. It's more likely there's something else which has caused memory usage to grow so much.

Were those tests done using a dockerd which was just started and no other workloads? If that's not the case, it's quite likely starting and stopping containers had nothing to do with this issue.

Please enable debugging and fetch the pprof profile.

**Comment by fortinj66** [comment]
So I've enabled debugging for the dockerd-current daemon and I am able to pull a memory profile.

I get the following:
```
100  461k    0  461k    0     0  18.9M      0 --:--:-- --:--:-- --:--:-- 19.5M
[johnf@os-node-p01 ~]$  go tool pprof --inuse_space /usr/bin/dockerd-current heap.txt
Entering interactive mode (type "help" for commands)
(pprof) top5
2858.40MB of 2947.68MB total (96.97%)
Dropped 788 nodes (cum <= 14.74MB)
Showing top 5 nodes out of 27 (cum >= 44.50MB)
      flat  flat%   sum%        cum   cum%
 2756.37MB 93.51% 93.51%  2757.37MB 93.54%  io.copyBuffer
   76.03MB  2.58% 96.09%    76.03MB  2.58%  runtime.malg
   12.50MB  0.42% 96.51%    19.50MB  0.66%  github.com/docker/docker/pkg/ioutils.NewBytesPipe
       8MB  0.27% 96.78%    15.50MB  0.53%  github.com/docker/docker/container/stream.(*Config).CopyStreams
    5.50MB  0.19% 96.97%    44.50MB  1.51%  github.com/docker/docker/daemon.(*Daemon).ContainerExecStart
(pprof)

[johnf@os-node-p01 ~]$  go tool pprof --alloc_space /usr/bin/dockerd-current heap.txt
Entering interactive mode (type "help" for commands)
(pprof) top5
369.19GB of 503.58GB total (73.31%)
Dropped 698 nodes (cum <= 2.52GB)
Showing top 5 nodes out of 117 (cum >= 19.77GB)
      flat  flat%   sum%        cum   cum%
  158.85GB 31.54% 31.54%   158.85GB 31.54%  github.com/docker/docker/daemon/exec.(*Store).List
  117.19GB 23.27% 54.82%   117.19GB 23.27%  bytes.makeSlice
   45.55GB  9.05% 63.86%    51.64GB 10.26%  io/ioutil.readAll
   27.82GB  5.52% 69.39%    29.26GB  5.81%  io.copyBuffer
   19.77GB  3.93% 73.31%    19.77GB  3.93%  reflect.unsafe_NewArray
(pprof)
```
looks like io.copyBuffer is the heavy hitter...

--John

**Comment by unclejack** [comment]
@fortinj66: Thank you for not giving up on this and for providing the output. I'll investigate this shortly.

**Comment by fortinj66** [comment]
and here is an example for a node running for jsut a couple of hours.  This node is basically running Jenkins and Selenium nodes.  After about two days dockerd will use almost all of the memory on the server:

```
[root@os-node-s01 docker]#  go tool pprof --inuse_space $(which dockerd-current) heap.txt
Entering interactive mode (type "help" for commands)
(pprof) top10
178.82MB of 185.40MB total (96.45%)
Dropped 720 nodes (cum <= 0.93MB)
Showing top 10 nodes out of 70 (cum >= 1.02MB)
      flat  flat%   sum%        cum   cum%
  158.86MB 85.69% 85.69%   159.36MB 85.96%  io.copyBuffer
    5.50MB  2.97% 88.65%     5.50MB  2.97%  runtime.malg
    3.29MB  1.78% 90.43%     3.29MB  1.78%  github.com/docker/libnetwork/ipamutils.initGranularPredefinedNetworks
    2.58MB  1.39% 91.82%     2.58MB  1.39%  regexp.(*bitState).reset
    2.04MB  1.10% 92.92%     2.04MB  1.10%  github.com/docker/docker/daemon/graphdriver/devmapper.NewDeviceSet
    1.50MB  0.81% 93.73%     1.50MB  0.81%  github.com/docker/docker/container/stream.(*Config).CopyStreams.func3
    1.50MB  0.81% 94.54%     2.50MB  1.35%  github.com/docker/docker/pkg/ioutils.NewBytesPipe
    1.50MB  0.81% 95.35%     1.50MB  0.81%  github.com/docker/docker/pkg/ioutils.getBuffer.func1
    1.03MB  0.56% 95.90%     1.03MB  0.56%  github.com/docker/docker/pkg/pools.newBufioReaderPoolWithSize.func1
    1.02MB  0.55% 96.45%     1.02MB  0.55%  github.com/docker/docker/daemon/logger.(*Copier).copySrc
(pprof)
```

**Comment by unclejack** [comment]
@fortinj66: Can you upload the files as well, please?

**Comment by fortinj66** [comment]
This is from the first set of data.  If you need the other let me know...

[heap.txt](https://github.com/moby/moby/files/1038731/heap.txt)

**Comment by unclejack** [comment]
@fortinj66: What version of Docker were you running? Was it installed from the CentOS repositories?

**Comment by fortinj66** [comment]
I’m away from my desk but the version is the latest from the Centos7 repositories…

I can get you the exact version when I get back to my desk…

—John

**Comment by fortinj66** [comment]
Here you go:

```
[johnf@os-node-p01 ~]$ docker version
Client:
 Version:         1.12.6
 API version:     1.24
 Package version: docker-common-1.12.6-16.el7.centos.x86_64
 Go version:      go1.7.4
 Git commit:      3a094bd/1.12.6
 Built:           Fri Apr 14 13:46:13 2017
 OS/Arch:         linux/amd64

Server:
 Version:         1.12.6
 API version:     1.24
 Package version: docker-common-1.12.6-16.el7.centos.x86_64
 Go version:      go1.7.4
 Git commit:      3a094bd/1.12.6
 Built:           Fri Apr 14 13:46:13 2017
 OS/Arch:         linux/amd64
[johnf@os-node-p01 ~]$ rpm -qa | grep docker
docker-client-1.12.6-16.el7.centos.x86_64
docker-common-1.12.6-16.el7.centos.x86_64
docker-1.12.6-16.el7.centos.x86_64
origin-docker-excluder-1.4.1-1.el7.noarch
```

**Comment by fortinj66** [comment]
Well, I guess it isn't the "current" latest and greatest...  But I did update within the last 2 weeks...

Todays latest is: docker.x86_64 2:1.12.6-16.el7.  Minor numbers are the same...

**Comment by fortinj66** [comment]
I'm going to update on the node that I sent the data on.  We'll see what happens...  I should know fairly quickly

**Comment by fortinj66** [comment]
Here you go...  I may have gotten the two heap.txt file reversed...

[heap.txt](https://github.com/moby/moby/files/1039004/heap.txt)

**Comment by unclejack** [comment]
@fortinj66: I've tried to investigate the issue. I'll have to look further into this. So far, I've only figured out that the profile isn't matching the binaries from the CentOS packages for some reason.

**Comment by thaJeztah** [comment]
Version seems to be from this commit in their repository https://github.com/projectatomic/docker/commit/3a094bd

**Comment by fortinj66** [comment]
so should I take a stab at taking the latest and greatest Centos source, apply the patch and gice it a whirl?

The current version has dockerd at over 5Gig of memory after 3 days...

**Comment by cpuguy83** [comment]
@fortinj66 That'd be really helpful. Though it looks like we have a minor issue with some failing tests with the changes.

**Comment by fortinj66** [comment]
Unfortunately I was not successful in building/running a patched version.  i was finally able to get it to build (after having to patch several build files) the binary itself would not start:

"Error starting daemon: error initializing graphdriver: devmapper: Deferred removal can not be enabled as libdm does not support it"

I'm guessing that there are additional build arguments that need to be passed but I can't seem to find them...

**Comment by fortinj66** [comment]
Well, it looks like we are seeing this in our swarm environment.  This is a 1.12.3 setup of vanilla docker )(yes, I know...  old as dirt...)  dockerd was up to almost 4 Gig.

[johnf@shop-docker-p01 yum.repos.d]$ sudo docker info
Containers: 8
 Running: 8
 Paused: 0
 Stopped: 0
Images: 11
Server Version: 1.12.3
Storage Driver: devicemapper
 Pool Name: docker--vg-thinpool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 4.72 GB
 Data Space Total: 25.47 GB
 Data Space Available: 20.75 GB
 Metadata Space Used: 1.253 MB
 Metadata Space Total: 234.9 MB
 Metadata Space Available: 233.6 MB
 Thin Pool Minimum Free Space: 2.546 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: active
 NodeID: 5jadtmx5a3ydf0hhzei0lu1kc
 Is Manager: true
 ClusterID: e75pr3vpz1wro5se6y9fvawlt
 Managers: 3
 Nodes: 6
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.64.135
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.36.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 15.67 GiB
Name: shop-docker-p01.corp.shop.com
ID: XKB7:QXDK:DJ5T:U5QL:SDWM:4Z3T:RG53:YOPF:BFWA:GIB4:OW7R:FIV6
Docker Root Dir: /var/lib/docker/100000.100000
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8

**Comment by baillieny** [comment]
I'm seeing the same problem on RacnherOS using the centos console

docker info
Containers: 120
Running: 98
Paused: 0
Stopped: 22
Images: 176
Server Version: 17.03.1-ce
Storage Driver: overlay2
Backing Filesystem: extfs
Supports d_type: true
Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
Volume: local
Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: N/A (expected: 4ab9917febca54791c5f071a9d1f404867857fcc)
runc version: N/A (expected: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe)
init version: N/A (expected: 949e6facb77383876aeff8a6944dde66b3089574)
Security Options:
seccomp
Profile: default
Kernel Version: 4.9.21-rancher
Operating System: RancherOS v1.0.0
OSType: linux
Architecture: x86_64
CPUs: 10
Total Memory: 19.61 GiB
Name: INTDC1MMAX14
ID: BJAD:JZTL:WXKZ:6SJH:AMOL:4457:NKS7:QOJL:MF3W:K4BY:FYEB:I2QM
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
File Descriptors: 10655
Goroutines: 30527
System Time: 2017-06-13T00:20:07.899030681Z
EventsListeners: 10032
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
127.0.0.0/8
Registry Mirrors:
https://docker-mirror.docker.tattsgroup.com/
Live Restore Enabled: false

Using centos console
After a 3-4 days dockerd consumes a large amount of memory

top - 00:22:07 up 6 days, 21:38, 0 users, load average: 220.88, 241.41, 257.26
Tasks: 472 total, 5 running, 467 sleeping, 0 stopped, 0 zombie
%Cpu(s): 1.2 us, 29.5 sy, 0.0 ni, 0.2 id, 68.7 wa, 0.0 hi, 0.5 si, 0.0 st
KiB Mem : 20562184 total, 166300 free, 17294900 used, 3100984 buff/cache
KiB Swap: 4194300 total, 0 free, 4194300 used. 224816 avail Mem

PID USER PR NI VIRT RES SHR S %CPU %MEM TIME+ COMMAND
1530 root 20 0 9.983g 3.101g 0 S 14.0 15.8 499:07.33 dockerd
7864 105 20 0 8567252 2.055g 0 S 2.0 10.5 8:01.87 java
11968 1000 20 0 11.940g 1.852g 0 S 6.0 9.4 23:01.34 java
17103 1000 20 0 8133656 1.280g 0 S 4.8 6.5 20:14.58 java
1141 root 20 0 2336364 1.061g 0 D 5.0 5.4 879:55.07 node
7520 1000 20 0 9394176 920316 0 S 2.8 4.5 276:17.32 java
21130 1000 20 0 7534088 804656 0 S 1.3 3.9 391:34.42 java
15927 105 20 0 9351724 499208 0 S 3.5 2.4 2783:43 java

Docker debug
go tool pprof --inuse_space /usr/bin/dockerd heap.txt
Entering interactive mode (type "help" for commands)
(pprof) top5
2519.65MB of 2638.57MB total (95.49%)
Dropped 1584 nodes (cum <= 13.19MB)
Showing top 5 nodes out of 59 (cum >= 31.01MB)
flat flat% sum% cum cum%
2217.84MB 84.05% 84.05% 2217.84MB 84.05% bytes.makeSlice
157.41MB 5.97% 90.02% 157.41MB 5.97% github.com/docker/docker/pkg/pubsub.(*Publisher).SubscribeTopic
66.21MB 2.51% 92.53% 66.21MB 2.51% net/http.newBufioWriterSize
47.18MB 1.79% 94.32% 47.18MB 1.79% net/http.newBufioReader
31.01MB 1.18% 95.49% 31.01MB 1.18% runtime.malg

go tool pprof --alloc_space /usr/bin/dockerd heap.txt
Entering interactive mode (type "help" for commands)
(pprof) top5
364.97GB of 942.35GB total (38.73%)
Dropped 1481 nodes (cum <= 4.71GB)
Showing top 5 nodes out of 162 (cum >= 49.04GB)
flat flat% sum% cum cum%
122.48GB 13.00% 13.00% 123.09GB 13.06% io/ioutil.readAll
89.57GB 9.50% 22.50% 89.57GB 9.50% github.com/docker/docker/pkg/pubsub.(*Publisher).SubscribeTopic
52.93GB 5.62% 28.12% 66.16GB 7.02% github.com/docker/docker/daemon/events.(*Filter).matchEvent
50.95GB 5.41% 33.53% 50.95GB 5.41% reflect.unsafe_New
49.04GB 5.20% 38.73% 49.04GB 5.20% reflect.unsafe_NewArray

**Comment by unclejack** [comment]
@baillieny: Can you provide the heap.txt file and the dockerd binary, please?

**Comment by baillieny** [comment]
@unclejack heap is attached to my comment above. How do I get the dockerd binary

**Comment by unclejack** [comment]
@baillieny: That's ok. I'll try to obtain it from a copy of Rancher to investigate. I've missed the heap.txt in your post, sorry.

**Comment by SvenDowideit** [comment]
@unclejack we use the plain, unmodified docker-ce binaries when they're available (ie, we only build arm64 ourselves.)

**Comment by SvenDowideit** [comment]
@unclejack why is this issue closed?

**Comment by thaJeztah** [comment]
@SvenDowideit it was closed by https://github.com/moby/moby/issues/33472, but that's 17.06

**Comment by panpan0000** [comment]
Hi，Guys, I found another similar memory leak about **dockerd** , and easy to be reproduced in 17.0.3.2.
create so containers( I've 100+)
generate a lot of error log into docker daemon:
`while true; do docker ps |awk '{print $1}'|xargs -I {} docker exec {} any_invalid_command;  done`
then you will find out that dockerd RSS memory bump from 120MB to 1GB in one hour.
I suspected that huge logs will cause this problem. but no prove (I tried keep doing "kill -SUGHUP" to dockerd, it increased memory as well. )

By the pprof analytic : `container.AttachStreams.func2` -> `io.Copy` eats up of the memory. 

But I guess it was  fixed by #33472
because when I upgrade to 17.06.2, issue gone ( memory bumped still, but just a few MB .)
